src/passes/string-encoding/Routines.cpp

```diff
@@ -0,0 +1,63 @@
+#include "omvll/passes/string-encoding/Routines.h"
+#include <cassert>
+
+namespace {
+
+const char *DecodeRoutines[] = {
+  R"delim(
+    void decode(char *out, char *in, unsigned long long key, int size) {
+      unsigned char *raw_key = (unsigned char*)(&key);
+      for (int i = 0; i < size; ++i) {
+        out[i] = in[i] ^ raw_key[i % sizeof(key)];
+      }
+    }
+  )delim",
+  R"delim(
+    void decode(char *out, char *in, unsigned long long key, int size) {
+      unsigned char *raw_key = (unsigned char*)(&key);
+      for (int i = 0; i < size; ++i) {
+        out[i] = in[i] ^ raw_key[i % sizeof(key)] ^ i;
+      }
+    }
+  )delim",
+};
+
+omvll::EncRoutineFn *EncodeRoutines[] = {
+    [](char *out, const char *in, unsigned long long key, int size) {
+      unsigned char *raw_key = (unsigned char *)(&key);
+      for (int i = 0; i < size; ++i) {
+        out[i] = in[i] ^ raw_key[i % sizeof(key)];
+      }
+    },
+    [](char *out, const char *in, unsigned long long key, int size) {
+      unsigned char *raw_key = (unsigned char *)(&key);
+      for (int i = 0; i < size; ++i) {
+        out[i] = in[i] ^ raw_key[i % sizeof(key)] ^ i;
+      }
+    }};
+
+} // namespace
+
+namespace omvll {
+
+template <typename T, unsigned N>
+static constexpr unsigned arraySize(const T (&)[N]) noexcept {
+  return N;
+}
+
+// Encode and decode functions must match in pairs
+static_assert(arraySize(EncodeRoutines) == arraySize(DecodeRoutines));
```

Comment:
How about something like:
```cpp
struct RoutinePair {
    EncRoutineFn *Encode;
    const char *Decode;
};
static constexpr std::array<RoutinePair, 2> Routines = { ... };
```
Then you could have the following without the need of the static_assert and arraySize template: 
```cpp
unsigned getNumRoutines() { return std::size(Routines); }
EncRoutineFn *getEncodeRoutine(unsigned Idx) { return Routines[Idx].Encode; }
const char *getDecodeRoutine(unsigned Idx) { return Routines[Idx].Decode; }
```

Reply:
This also affects the initialization in the cpp. Can we consider it as an improvement in another PR?

Reply:
Sounds good, let's leave this for a future improvement.